README.md
# Isaac Debug Helper

This tool is aimed at anyone who wishes to create mods for *The Binding of Isaac*. It's a log reader: reads the log files directly and displays the content in a user-friendly graphical user interface. 
The goal of this project is to create a complete software suite for modding *The Binding of Isaac*.


License: GNU GPL 3.0

## Screenshot
![Isaac Debug Helper 2.1.0 on Windows][pic]

[pic]: https://github.com/teamdodo/isaac-debug-helper/blob/master/pics/Isaac_Debug_Helper_2.1.0_720x480.jpg "Isaac Debug Helper 2.1.0 on Windows"


## Requirements :

### Executables
Executables should work out of the box. If they don't, tell us.

### Python script
 - Python 3.x 
 - On Linux you might have to install manually tcl/tk (tkinter) for Python if it's not already installed.
 
#### Optional

 - [psutil] (pip install psutil) - Needed for autoreload feature.

 All other libraries are standard libraries.

## Tested on:
* Python 3.6.4 on Windows 7
* Python 3.5.1-3 with python3-tk 3.5.1-1 on Lubuntu 16.04 LTS

# Features :

 - *Start* and *Stop* menu buttons, to load or reload the script when debugging
 - Color highlighting errors, warnings, informations and debug information
 - A GUI designed for ease of use, with most of the window dedicated to the output
 - The log file is automatically found, no need for any configuration, it works out of the box
 - Fully designed to be cross-platform (at least as long as Python support it)
 - Dynamic LUA memory usage while the debugger is running, with a progress bar to make the logs even more readable
 - Scrollbar for the text widget
 - Option Toolbox to set up custom tags and filters for the debugger, changing colors etc
 - Auto reload feature if *The Binding of Isaac* is restarted
 - Store preferences in a local config file
 - Slider for maximum LUA memory (from 1KB to 10KB) in the options
 - About menu

# Changelog
See [CHANGELOG.md]
# About the project :

A modification by [Team Dodo] of [Isaac Modding Helper Tool]. Original project's credits:
 - Lead Developer: Dogeek
 - Debugger Idea: Krayz
 - Layout Design: Dogeek

[CHANGELOG.md]: /CHANGELOG.md
[Isaac Modding Helper Tool]: https://github.com/Dogeek/isaac-debug-helper
[psutil]: https://github.com/giampaolo/psutil
[Team Dodo]: https://github.com/teamdodo/


CHANGELOG.md
# Changelog
All notable changes to *Isaac Debug Helper* project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
None

## 2.1.0 - 2018-02-26

### Added 
- Check for updates (check if a new version is available for download).
- Exceptions are now reported as errors.

## 2.0.0 - 2018-02-11

### Added 
- GUI: debugger status indicator.
- GUI: Lua memory usage counter on the progress bar.
- *Clear* text function.
- Button to reset options to default.
- Version number and Team Dodo credits in *About*.
- Handle config.cfg not found or damaged.
- Handle psutil not found.

### Changed
- Added function called on WM_DELETE_WINDOW of Options window.
- Options are now saved in config.cfg both when *Ok* button is clicked and when Isaac debug helper is closedonOkButton and on_closing are triggered.
- Options are **not** saved when *Options* window is open, user changes settings then closes parent window. In this case user is supposed to discard changes.
### Removed
- Old #TODOs in code

### Fixed
- GUI: adjusted buttons size in Options.
- "SyntaxError: encoding problem: utf8" when launching by shell or double-click.
	* Changed encoding name to utf-8 in .py files.
- Options not saving, Options not opening again after being closed.
	* It looked like a problem with global variables: solved merging options.py and main.py. Brutal solution but it works.
- "DebugFrame has no object log_f" when clicking on *Stop* but the debugger has never been started before.
	
## Prior to fork - up to 2018-02-11
See [Isaac Modding Helper Tool] by Dogeek

[Isaac Modding Helper Tool]: https://github.com/Dogeek/isaac-debug-helper/


debug.py
import tkinter as tk
import tkinter.ttk as ttk
import tkinter.messagebox as messagebox
from constants import *
from functions import *

class DebugFrame(tk.Frame):
	def __init__(self, master=None):
		tk.Frame.__init__(self,master)
		self.master = master
		self.started = False
		self.log_path = get_log_path()
		self.oldline = "  "
		self.tags = TAGS
		self.max_mem = MAX_LUA_MEMORY*1024
		self.mem = 0
		self.init_layout()
		pass

	def init_layout(self):	
		self.labelStatus = tk.Label(self, text="Debugger status: stopped", foreground="red")
		self.labelStatus.pack(side=tk.TOP, fill=tk.X)
		
		self.scrollbar = tk.Scrollbar(self)
		self.scrollbar.pack(side=tk.RIGHT, fill=tk.Y)

		self.progressbar = ttk.Progressbar(self, orient="horizontal", length=410, mode="determinate")
		self.progressbar.pack(side=tk.BOTTOM, fill=tk.X)
		self.progressbar["value"] = 0
		self.progressbar["maximum"] = self.max_mem
		
		
		self.label = tk.Label(self, text=("LUA", "memory","usage",":", self.mem, "of", self.max_mem, "Bytes"))
		self.label.pack(side=tk.BOTTOM,in_=self.progressbar)

		self.output = tk.Text(self)
		self.tag_config()
		self.output.config(font="sans 10", width=200, height=60, yscrollcommand=self.scrollbar.set)
		self.output.pack(side=tk.LEFT, fill=tk.BOTH)
		self.scrollbar.config(command=self.output.yview)
		self.readfile()
		pass

	def tag_config(self):
		for key, value in self.tags.items():
			self.output.tag_config(key, foreground=value)
		pass

	def readfile(self):
		if self.started:
			self.tags = TAGS
			self.tag_config()
			tmp = self.log_f.readline().lower()
			if self.oldline != tmp: #display spam only once@FileLoad
				self.output.config(state=tk.NORMAL)
				if "err" in tmp or "exception" in tmp and not "overlayeffect" in tmp and not "animation" in tmp: #Error filter to display
					self.output.insert(tk.END, tmp, "error")
				elif "lua mem usage" in tmp:
					mem_txt = tmp.split("lua mem usage: ")[1]
					kb = mem_txt.split(" kb")[0]
					by = mem_txt.split(" and ")[1].split(" bytes")[0]
					try:
						kb = int(kb)
						by = int(by)
					except ValueError as e:
						print(e)
						exit(1)
					self.progressbar["value"] = kb*1024+by
					
					self.mem = kb*1024+by
					self.label.config(text=("LUA", "memory","usage",":", self.mem, "of", self.max_mem, "Bytes"))
				elif "lua" in tmp and not "debug" in tmp:
					self.output.insert(tk.END, tmp, "info")
				elif "warn" in tmp:
					self.output.insert(tk.END, tmp, "warning")
				elif "lua debug" in tmp:
					self.output.insert(tk.END, tmp, "luadebug")
				self.oldline = tmp
			self.output.see(tk.END)
			self.update_idletasks()
			self.after(5, self.readfile)
		pass
	def start(self):
		self.labelStatus.config(text="Debugger status: started", foreground="green")
		self.log_f = open(self.log_path, "r")
		self.started = True
		self.readfile()
		pass
	def stop(self):
		try: #In case of click on "Stop" when debugger has never been started. Otherwise raises exception "DebugFrame has no object log_f"
			self.log_f.close()
			self.labelStatus.config(text="Debugger status: stopped", foreground="red")
			self.started = False
		except AttributeError:
			pass
		pass
	def clear(self):
		if not self.started :
			self.output.delete(1.0,"end")
			self.progressbar["value"] = 0
			self.mem = 0
			self.label.config(text=("LUA", "memory","usage",":", self.mem, "of", self.max_mem, "Bytes"))
		else:
			messagebox.showinfo(title="Warning", icon="warning", message="Debugger is running! You have to stop it before you can clear the text.")
		pass

	def auto_reload(self):
		global AUTO_RELOAD
		if AUTO_RELOAD:
			if is_isaac_running() and not self.started:
				self.start()
			elif not is_isaac_running() and self.started:
				self.stop()
	pass


update.py
import configparser
import os
import tkinter.messagebox as messagebox
from urllib.request import urlretrieve as download
import webbrowser

from constants import VERSION

class updateCheckClass():
	def update():
		updateCheck = messagebox.askyesno(title="Update check", message="Check for updates?")
		if updateCheck == True:
			try:
				download("http://raw.githubusercontent.com/teamdodo/isaac-debug-helper/master/update/update.cfg" , "update.cfg")
				config = configparser.ConfigParser()
				config.read("update.cfg")
				latestVersion = config["UPDATE"]["latestVersion"]
				latestVersionUrl = config["UPDATE"]["latestVersionUrl"]
				os.remove("update.cfg")
				if VERSION != latestVersion:
					updateMessage = "Un update is avaliable, do you want to download it? This will launch your default browser and open a link to download the update"
					updateDetail = "Your version: {0} \nLatest version: {1}".format(VERSION,latestVersion)
					updateDownload = messagebox.askyesno(title="Update avaliable", message=updateMessage, detail=updateDetail)
					if updateDownload == True:
						webbrowser.open(latestVersionUrl, new=1, autoraise=True)
				if VERSION == latestVersion:
					messagebox.showinfo(title="No update avaliable", message="No updates avaliable. You already have the latest version")
			except urllib.error.HTTPError:
				messagebox.showinfo(title="Error", icon="error", message="Error while downloading update info. Try again later")
			except KeyError or IOError:
				messagebox.showinfo(title="Error", icon="error", message="Error while reading update info. Try again later")
			except OSError:
				messagebox.showinfo(title="Error", icon="error", message="Error while deleting update info update.cfg. Try to delete it manually")
			except webbrowser.Error:
				messagebox.showinfo(title="Error", icon="error", message="Error while opening your browser. Try again later or visit manually {0}".format(latestVersionUrl))
			except:
				messagebox.showinfo(title="Error", icon="error", message="Unmanaged error while checking for updates. Try again later. \nIf this persists contact the developers.")
	pass


options.py
#-*- encoding:utf8 -*-
#!/usr/bin/python3

import tkinter as tk
from tkinter import colorchooser
from constants import *

class OptionGUI(tk.Toplevel): #TODO : make an option window to pick colors for error highlighting, add new filters
	def __init__(self, master = None):
		super(OptionGUI, self).__init__()
		global AUTO_RELOAD
		global MAX_LUA_MEMORY
		self.master = master
		self.master.title = "Options"
		self.tags = TAGS
		self.maxmem_var = tk.IntVar()
		self.maxmem_var.set(MAX_LUA_MEMORY)
		self.checkbutton_var = tk.IntVar()
		if AUTO_RELOAD:
			self.checkbutton_var.set(1)
		else:
			self.checkbutton_var.set(0)
		self.labels = []
		self.buttons = []
		self.config()
	def config(self):
		i=0
		for key, value in self.tags.items():
			self.labels.append(tk.Label(self, text=key))
			but = tk.Button(self, text=" ", width=10, background=value, command=lambda x=i: self.open_color_picker(x))
			self.buttons.append(but)
			i += 1
		for i in range(len(self.labels)):
			self.labels[i].grid(column=0, row=i)
			self.buttons[i].grid(column=1, row=i)
			self.buttons[i].config(command=lambda x=i:self.open_color_picker(x))
		self.checkbutton = tk.Checkbutton(self, variable=self.checkbutton_var, text="Auto-reload")
		self.checkbutton.grid(row=4, column=0, columnspan=2)
		tk.Label(self, text="Maximum LUA Memory (KB): ").grid(row=5, column=0)
		self.maxmem_scale = tk.Scale(self, variable=self.maxmem_var, orient="horizontal", from_=1, to=10*1024, command=self.scaleToGlobal)
		self.maxmem_scale.grid(row=5, column=1)

		self.ok_button = tk.Button(self, text="Ok", command=self.onOkButton)
		self.cancel_button = tk.Button(self, text="Cancel", command=self.onCancelButton)
		self.ok_button.grid(column=0, row=6)
		self.cancel_button.grid(column=1, row=6)
		self.loopCheck()

	def open_color_picker(self, i):
		color = colorchooser.askcolor()
		if color[1] is not None:
			color = color[1]
		else:
			color = self.buttons[i]["background"]
		self.buttons[i].config(background=color)
		self.tags[self.labels[i]["text"]] = color
	def onOkButton(self):
		global TAGS
		global OPTIONSOPEN
		TAGS=self.tags
		OPTIONSOPEN = False
		self.destroy()
	def onCancelButton(self):
		global OPTIONSOPEN
		OPTIONSOPEN = False
		self.destroy()
	def scaleToGlobal(self, arg):
		global MAX_LUA_MEMORY
		MAX_LUA_MEMORY = self.maxmem_var.get()
	def loopCheck(self):
		global AUTO_RELOAD
		if self.checkbutton_var.get():
			AUTO_RELOAD = True
		else:
			AUTO_RELOAD = False
		self.after(20, self.loopCheck)
	pass


constants.py
import configparser

OPTIONSOPEN = False
VERSION = "2.1.0" #Used to check for updates
try: 
	config = configparser.ConfigParser()
	config.read("config.cfg")
	TAGS = config["TAGS"]
	MAX_LUA_MEMORY = config["GENERAL"].getint("maxmem")
	AUTO_RELOAD = config["GENERAL"].getboolean("autoreload")
	GEOMETRY = config["GEOMETRY"]["x"] + "x" + config["GEOMETRY"]["y"]
except KeyError or IOError: 
	import tkinter.messagebox as messagebox
	messagebox.showinfo(title="Error", icon="error", message="Config file not found or damaged. Config has been set to default!")
	config = configparser.ConfigParser()
	config["TAGS"] = {'error': '#ff0080','warning': '#00FF00', 'info': '#0000FF', 'luadebug': '#000000'}
	config["GENERAL"] = {}
	config["GENERAL"]["maxmem"] = "3096"
	config["GENERAL"]["autoreload"] = "True"
	config["GEOMETRY"] = {}
	config["GEOMETRY"]["x"] = "1018"
	config["GEOMETRY"]["y"] = "882"
	with open('config.cfg', 'w') as configfile:
		config.write(configfile)
	config.read("config.cfg")
	TAGS = config["TAGS"]
	MAX_LUA_MEMORY = config["GENERAL"].getint("maxmem")
	AUTO_RELOAD = config["GENERAL"].getboolean("autoreload")
	GEOMETRY = config["GEOMETRY"]["x"] + "x" + config["GEOMETRY"]["y"]	


functions.py
import sys
from os.path import expanduser
#External libraries
try:
	import psutil
except ImportError:
	import tkinter.messagebox as messagebox
	messagebox.showinfo(title="Psutil not found", icon="error", message="Psutil not found. Autoreload will NOT work. \n Install it and launch the debugger again.")


def get_log_path():
	p = ""
	if sys.platform == "linux" or sys.platform == "linux2":
		p = expanduser("~/.local/share/binding of isaac afterbirth+/log.txt")
	elif sys.platform == "darwin":
		p = expanduser("~/Library/Application Support/Binding of Isaac Afterbirth+/log.txt")
	elif sys.platform == "win32":
		p = expanduser("~/Documents/My Games/binding of isaac afterbirth+/log.txt")
	return p

def is_isaac_running():
	try:
		for p in psutil.process_iter():
			try:
				if 'isaac-ng' in p.name():
					return True
			except psutil.Error:
				pass
		return False
	except  NameError: #NameError in case of psutil not found
		return False

def is_update_available(current_version):
	try:
		import requests
		from bs4 import BeautifulSoup
		import re
		from distutils.version import StrictVersion
	except ImportError:
		import tkinter.messagebox as messagebox
		messagebox.showerror("Error", message="You need BeautifulSoup4, distutils and requests installed for version checking to work.")

	url = "https://github.com/Dogeek/isaac-debug-helper/releases"
	page = requests.get(url)
	soup = BeautifulSoup(page.content, "html.parser")
	versions = [x for x in soup.find_all("span", class_="css-truncate-target")]
	github_version = versions[-1] if versions else "0.1"

	regex = re.compile(r"([\d.]+)")
	assert regex.match(current_version) is not None, "Current Version doesn't match the standardized version format (x.x.x.x)"
	assert regex.match(github_version) is not None, "Github Version doesn't match the standardized version format (x.x.x.x)"
	return StrictVersion(github_version)>StrictVersion(current_version)


old_code/hyperlink_text_message.py
import re
import webbrowser
from tkinter import *

class HyperlinkMessageBox(Toplevel):
    hyperlinkPattern = re.compile(r'<a href="(?P<address>.*?)">(?P<title>.*?)'
                                  '</a>')
    def __init__(self, master, title=None, message=None, **options):
        Toplevel.__init__(self, master)
        self.geometry("200x200")
        self.title(title or master.title())
        self.text = Text(self, wrap=WORD, bg=master.cget('bg'),
                         height=self.cget('height'))
        self._formatHyperLink(message)
        self.text.config(state=DISABLED)
        self.text.pack(side=TOP, fill=X)
        self.makeButtons()

    def makeButtons(self):
        Button(self, text="Ok",
               command=lambda *a, **k: self.destroy()).pack()

    def _formatHyperLink(self, message):
        text = self.text
        start = 0
        for index, match in enumerate(self.hyperlinkPattern.finditer(message)):
            groups = match.groupdict()
            text.insert("end", message[start: match.start()])
            #insert hyperlink tag here
            text.insert("end", groups['title'])
            text.tag_add(str(index),
                         "end-%dc" % (len(groups['title']) + 1),
                         "end",)
            text.tag_config(str(index),
                            foreground="blue",
                            underline=1)
            text.tag_bind(str(index),
                          "<Enter>",
                          lambda *a, **k: text.config(cursor="hand"))
            text.tag_bind(str(index),
                          "<Leave>",
                          lambda *a, **k: text.config(cursor="arrow"))
            text.tag_bind(str(index),
                          "<Button-1>",
                          self._callbackFactory(groups['address']))
            start = match.end()
        else:
            text.insert("end", message[start:])

    def _callbackFactory(self, url):
        return lambda *args, **kwargs: webbrowser.open(url)


if __name__ == "__main__":
    import re
    def dirsel(obj, search):
        for item in dir(obj):
            if re.search(search, item):
                print(item)

    root = Tk()
    h = HyperlinkMessageBox(root, "My App", 'Some message <a href="http://www.google.com">Google</a>.')
    root.mainloop()


main.py
#-*- encoding:utf-8 -*-
#!/usr/bin/python3

import tkinter as tk
import tkinter.ttk as ttk
import tkinter.messagebox as messagebox
import time
import configparser
from tkinter import colorchooser

from constants import *
from functions import *
from debug import DebugFrame
from update import *

class GUI(tk.Frame):
	def __init__(self, master=None):
		tk.Frame.__init__(self,master)
		self.master = master
		self.master.protocol("WM_DELETE_WINDOW", self.on_closing)
		self.init_layout()
		pass

	def init_layout(self):
		self.tabBar = ttk.Notebook(self)
		self.tabBar.pack(fill=tk.BOTH, padx=2, pady=3)
		self.tab_debugger = DebugFrame(self.tabBar)
		self.tabBar.add(self.tab_debugger, text="Debugger")

		self.menubar = tk.Menu(self)
		self.menubar.add_command(label="Start", command=self.tab_debugger.start)
		self.menubar.add_command(label="Stop", command=self.tab_debugger.stop)
		self.menubar.add_command(label="Clear", command=self.tab_debugger.clear)
		self.menubar.add_command(label="Options", command=self.open_options)
		self.menubar.add_command(label="Check for updates", command=updateCheckClass.update)
		self.menubar.add_command(label="About", command=self.about)
		self.master.config(menu=self.menubar)
		pass
	def on_closing(self):
		if not OPTIONSOPEN:
			config = configparser.ConfigParser()
			config["TAGS"] = TAGS
			config["GENERAL"] = {}
			config["GENERAL"]["maxmem"] = str(MAX_LUA_MEMORY)
			config["GENERAL"]["autoreload"] = str(AUTO_RELOAD)
			config["GEOMETRY"] = {}
			config["GEOMETRY"]["x"] = str(self.winfo_height())
			config["GEOMETRY"]["y"] = str(self.winfo_width())
			with open('config.cfg', 'w') as configfile:
				config.write(configfile)
		self.master.destroy()
	def about(self):
		message = """
		Isaac Debug Helper\n
		Version 2.1.0\n
		Fork by Team Dodo (based on software by Dogeek)\n
		For additional information, check out\n
		http://github.com/dogeek/isaac-debug-helper\n
		http://github.com/teamdodo/isaac-debug-helper\n
		License : GNU GPL 3.0
		"""
		messagebox.showinfo("About", message)
		pass

	def open_options(self):
		global OPTIONSOPEN
		if not OPTIONSOPEN:
			OPTIONSOPEN = True
			OptionGUI(master=self).mainloop()
	pass

class OptionGUI(tk.Toplevel): #TODO : make an option window to pick colors for error highlighting, add new filters
	def __init__(self, master = None):
		super(OptionGUI, self).__init__()
		global AUTO_RELOAD
		global MAX_LUA_MEMORY
		self.master = master
		self.title = "Options"
		self.protocol("WM_DELETE_WINDOW", self.on_closing)
		self.tags = TAGS
		self.maxmem_var = tk.IntVar()
		self.maxmem_var.set(MAX_LUA_MEMORY)
		self.checkbutton_var = tk.IntVar()
		if AUTO_RELOAD:
			self.checkbutton_var.set(1)
		else:
			self.checkbutton_var.set(0)
		self.labels = []
		self.buttons = []
		self.config()
	def config(self):
		i=0
		for key, value in self.tags.items():
			self.labels.append(tk.Label(self, text=key))
			but = tk.Button(self, text=" ", width=10, background=value, command=lambda x=i: self.open_color_picker(x))
			self.buttons.append(but)
			i += 1
		for i in range(len(self.labels)):
			self.labels[i].grid(column=0, row=i)
			self.buttons[i].grid(column=1, row=i)
			self.buttons[i].config(command=lambda x=i:self.open_color_picker(x))
		self.checkbutton = tk.Checkbutton(self, variable=self.checkbutton_var, text="Auto-reload")
		self.checkbutton.grid(row=4, column=0, columnspan=2)
		tk.Label(self, text="Maximum LUA Memory (KB): ").grid(row=5, column=0)
		self.maxmem_scale = tk.Scale(self, variable=self.maxmem_var, orient="horizontal", from_=1, to=10*1024, command=self.scaleToGlobal)
		self.maxmem_scale.grid(row=5, column=1)

		self.reset_button = tk.Button(self, text="Reset to default", command=self.onResetButton)
		self.ok_button = tk.Button(self, text="Ok", command=self.onOkButton)
		self.cancel_button = tk.Button(self, text="Cancel", command=self.onCancelButton)
		self.reset_button.grid(row=6, columnspan=2, sticky="ew")
		self.ok_button.grid(column=0, row=7, sticky="ew")
		self.cancel_button.grid(column=1, row=7, sticky="ew")
		self.loopCheck()
	
	def on_closing(self):
		global OPTIONSOPEN
		OPTIONSOPEN = False
		self.destroy()	
	def open_color_picker(self, i):
		color = colorchooser.askcolor()
		if color[1] is not None:
			color = color[1]
		else:
			color = self.buttons[i]["background"]
		self.buttons[i].config(background=color)
		self.tags[self.labels[i]["text"]] = color
	def onOkButton(self):
		global TAGS
		global OPTIONSOPEN
		TAGS=self.tags
		OPTIONSOPEN = False
		
		config = configparser.ConfigParser()
		config["TAGS"] = TAGS
		config["GENERAL"] = {}
		config["GENERAL"]["maxmem"] = str(MAX_LUA_MEMORY)
		config["GENERAL"]["autoreload"] = str(AUTO_RELOAD)
		config["GEOMETRY"] = {}
		config["GEOMETRY"]["x"] = str(self.master.winfo_height())
		config["GEOMETRY"]["y"] = str(self.master.winfo_width())
		with open('config.cfg', 'w') as configfile:
			config.write(configfile)
		self.destroy()
	def onCancelButton(self):
		global OPTIONSOPEN
		OPTIONSOPEN = False
		self.destroy()
	def onResetButton(self):
		resetConfirm = messagebox.askyesno(title="Reset config", icon="warning", message="This will reset options to default AND CLOSE THE DEBUGGER. \n Are you sure?")
		if resetConfirm == True:
			config = configparser.ConfigParser()
			config["TAGS"] = {'error': '#ff0080','warning': '#00FF00', 'info': '#0000FF', 'luadebug': '#000000'}
			config["GENERAL"] = {}
			config["GENERAL"]["maxmem"] = "3096"
			config["GENERAL"]["autoreload"] = "True"
			config["GEOMETRY"] = {}
			config["GEOMETRY"]["x"] = "1018"
			config["GEOMETRY"]["y"] = "882"
			with open('config.cfg', 'w') as configfile:
				config.write(configfile)
			sys.exit()
		else:
			global OPTIONSOPEN
			OPTIONSOPEN = False
			self.destroy()
	def scaleToGlobal(self, arg):
		global MAX_LUA_MEMORY
		MAX_LUA_MEMORY = self.maxmem_var.get()
	def loopCheck(self):
		global AUTO_RELOAD
		if self.checkbutton_var.get():
			AUTO_RELOAD = True
		else:
			AUTO_RELOAD = False
		self.after(20, self.loopCheck)
	pass
	
	
if __name__ == "__main__":
	root = tk.Tk()
	root.title("Isaac Debug Helper")
	root.geometry(GEOMETRY)
	gui = GUI(root)
	gui.pack()
	gui.mainloop()

